File: NeetCode/Linked List/AddTwoNumbers.java
public class AddTwoNumbers {
    
    public static class ListNode {
        int value;
        ListNode next;
        ListNode() {}
        ListNode(int value) {
            this.value = value;
        }
        ListNode(int value, ListNode next) {
            this.value = value;
            this.next = next;
        }
    }

    public static ListNode addTwoNumbers(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode();
        ListNode current = dummy;

        int carry = 0;
        while(list1 != null || list2 != null || carry != 0) {                   //! Time Complexity = O(n)
            int v1 = (list1 != null) ? list1.value : 0; 
            int v2 = (list2 != null) ? list2.value : 0;

            int sum = v1 + v2 + carry;
            carry = sum / 10;
            sum = sum % 10;
            current.next = new ListNode(sum);

            current = current.next;
            list1 = (list1 != null) ? list1.next : null;
            list2 = (list2 != null) ? list2.next : null;
        }

        return dummy.next;
    }                                                                           //! Total Time Complexity = O(n)

    public static void main(String[] args) {
        int[] values = {9};
        ListNode list1 = new ListNode(values[0]);
        ListNode current = list1;

        for(int i = 1;i < values.length;i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }

        int[] value = {1, 9, 9, 9, 9, 9, 9, 9, 9, 9};
        ListNode list2 = new ListNode(value[0]);
        current = list2;

        for(int i = 1;i < value.length;i++) {
            current.next = new ListNode(value[i]);
            current = current.next;
        }

        addTwoNumbers(list1, list2);
    }
}

File: NeetCode/Linked List/AddTwoNumbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummy, sum := new(ListNode), 0
    current := dummy

    for ;l1 != nil || l2 != nil || sum != 0; current = current.Next {
        
        if l1 != nil {
            sum += l1.Val
            l1 = l1.Next
        }
        
        if l2 != nil {
            sum += l2.Val
            l2 = l2.Next
        } 

        current.Next = &ListNode{Val: sum%10}
        sum /= 10
    }

    return dummy.Next;
}
